File: reverseLinkedList/reverseLinkedList.py
#!/usr/bin/env python3

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def print(self):
        l = self
        while(l):
            print(l.val)
            l = l.next

class Solution:
    def reverseList(self, head):
        if head == None:
            return None

        start = None
        lst = start

        current = head

        while(current):
            if(not current == None):
                lst = ListNode(current.val, lst)
                current = current.next
        
        return lst
    

if __name__ == "__main__":
    l = ListNode(5, ListNode(4, ListNode(3, ListNode(2, ListNode(1, None)))))
    s = Solution()
    m = s.reverseList(l)
    print("\n\n")
    m.print()

File: reverseLinkedList/reverseLinkedList.go
package main

import "fmt"

/**
 * Definition for singly-linked list.
 */
 type ListNode struct {
    Val int
    Next *ListNode
}

// print out a linked list
func (l *ListNode) print() {
	for l != nil {
		fmt.Println(l.Val)
		l = l.Next
	}
}

// constructLinkedList
// Takes O(n) time in the size of the list
func constructLinkedList(lst []int) *ListNode {
	l := &ListNode{lst[0], nil}
	c := l
	for _, v := range lst[1:] {
		l.Next = &ListNode{v, nil}
		l = l.Next
	}
	return c
}

// reverseList
func reverseList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	var lst *ListNode
	current := head

	for current != nil {
		lst = &ListNode{current.Val, lst}
		current = current.Next
	}
	
	start := lst

	return start
}

func main() {
	x := [...]int{2,3,4,1,8,4,5}

	lx := constructLinkedList(x[:])

	y := reverseList(lx)
	y.print()
}